c++/src/kj/memory.h

```diff
@@ -168,6 +170,53 @@ class NullDisposer: public Disposer {
   void disposeImpl(void* pointer) const override {}
 };
 
+
+// =======================================================================================
+// Ptr Counters
+
+#ifdef KJ_DEBUG
+#define KJ_ASSERT_PTR_COUNTERS
+// When defined, keeps track of active Ptr<T> instances and asserts validity of their ownership
+#endif
+
+namespace _ {
+#ifdef KJ_ASSERT_PTR_COUNTERS
+class AtomicPtrCounter {
+// AtomicPtrCounter uses atomic operations to keep track of active pointers.
+// 
+// Thread safety notes:
+// - every dec() observes the effect of corresponding inc() because of release/consume order
+// - when assertEmpty() is called with balanced dec/inc calls, all releases were consumed, and
+//   additional load with consume memory order will see the correct last value.
+// - when assertEmpty() is called with unbalanced dec/inc calls int will observe the result of 
+//   one of the unbalanced inc(), which is always > 0.
+
+public:
+  void dec() {
+    size_t prevCount = count.fetch_sub(1, std::memory_order_consume);
+    KJ_IREQUIRE(prevCount > 0, "unbalanced inc/dec");
```

Comment:
Should this crash instead of throw on failure? Since it suggests we're going to hit UB shortly anyway.

Reply:
yes, sure, but I wasn't able to find KJ_CRASH or something. Should I use abort() (and print to stderr)?

Reply:
You need to define some out-of-line function that crashes, then call it.

Probably the function should be `noexcept`, and then it should `KJ_FAIL_REQUIRE`. This produces a much better error report than `abort()`.

I suppose this pattern has come up enough to warrant adding macros for it, but that seems like a separate change...